dbms/src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -303,81 +314,103 @@ InterpreterSelectQuery::InterpreterSelectQuery(
+                if (is_subquery)
+                    interpreter_subquery = std::make_unique<InterpreterSelectWithUnionQuery>(
+                            table_expression,
+                            getSubqueryContext(context),
+                            options.subquery(),
+                            required_columns);
+            }
         }
-    }
-
-    if (interpreter_subquery)
-    {
-        /// If there is an aggregation in the outer query, WITH TOTALS is ignored in the subquery.
-        if (query_analyzer->hasAggregation())
-            interpreter_subquery->ignoreWithTotals();
-    }
 
-    required_columns = syntax_analyzer_result->requiredSourceColumns();
+        if (interpreter_subquery)
+        {
+            /// If there is an aggregation in the outer query, WITH TOTALS is ignored in the subquery.
+            if (query_analyzer->hasAggregation())
+                interpreter_subquery->ignoreWithTotals();
+        }
 
-    if (storage)
-    {
-        source_header = storage->getSampleBlockForColumns(required_columns);
+        required_columns = syntax_analyzer_result->requiredSourceColumns();
 
-        /// Fix source_header for filter actions.
-        if (context.hasUserProperty(storage->getDatabaseName(), storage->getTableName(), "filter"))
+        if (storage)
         {
-            filter_info = std::make_shared<FilterInfo>();
-            filter_info->column_name = generateFilterActions(filter_info->actions, storage, context, required_columns);
-            source_header = storage->getSampleBlockForColumns(filter_info->actions->getRequiredColumns());
+            source_header = storage->getSampleBlockForColumns(required_columns);
+
+            /// Fix source_header for filter actions.
+            if (context.hasUserProperty(storage->getDatabaseName(), storage->getTableName(), "filter"))
+            {
+                filter_info = std::make_shared<FilterInfo>();
+                filter_info->column_name = generateFilterActions(filter_info->actions, storage, context, required_columns);
+                source_header = storage->getSampleBlockForColumns(filter_info->actions->getRequiredColumns());
+            }
         }
-    }
 
-    /// Calculate structure of the result.
-    result_header = getSampleBlockImpl();
-    for (auto & col : result_header)
+        /// Calculate structure of the result.
+        result_header = getSampleBlockImpl();
+    };
+
+    analyze();
+
+    bool need_analyze_again = false;
+    if (analysis_result.prewhere_constant_filter_description.always_false || analysis_result.prewhere_constant_filter_description.always_true)
     {
-        if (!col.column)
-            col.column = col.type->createColumn();
-        else if (isColumnConst(*col.column) && !col.column->empty())
-            col.column = col.column->cloneEmpty();
+        auto constant = std::make_shared<ASTLiteral>(0u);
+        constant->alias = query.prewhere()->getColumnName();
```

Comment:
I don't think it's a good idea to make such aliases. It's quite OK just to rewrite expression with value, but there could be lots of troubles with alias collisions (at many stages of query processing). It's better to use autogenerated names for aliases as seldom as possible.

Reply:
But it just tries to reuse the old column name for predicates. I don't see how it can lead to collisions.

Reply:
And these names are bad. There'we plans to remove all the logic dependent on columns names outside of SyntaxAnalyzer. If we continue to use these bad names we never could make them good. In example we need CTE: ```WITH (some) as x``` which affects all the subqueries. And then we have ```some``` in subquery. The named items are not enough to make this. The same trash would happan if we have ```some``` in left and right part of not-INNER JOIN.

Even column names and aliases are not good to make these things. Autogenetared names are even worse. Cause expressions make temporary columns with such names. And we had some troubles there expression is out of expression chain but has the same temporary column name.

Reply:
> Autogenetared names are even worse. 

Now I'm confused. So what's the plan here? Perhaps some new naming infrastructure is under going now and this pr should be refactored after?

Reply:
I don't know a good solution. But my opinion is we could use Semantics (e.g. IdentifierSemantic, AsteriskSemantic). We could leave AST item as is but attach Semantic to it. Then when we decide to make expression for this AST item we could look at Semantic and chouse if it's const or some function.

Reply:
Ah, that makes sense, and the analyze output will be the same as before. Perhaps a PredicateSemantic then